File: Algo/LeadToOffer/17_SubTree.cpp
// Little bit hard. Think about it.
class Solution {
public:

    bool SameSubtree(TreeNode* pRoot1, TreeNode* pRoot2) {
        if (!pRoot2) return true;
        if (!pRoot1) return false;

        if (pRoot1->val != pRoot2->val) return false;

        return SameSubtree(pRoot1->left, pRoot2->left) && SameSubtree(pRoot1->right, pRoot2->right);
    }
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        if(!pRoot1 || !pRoot2) return false;
        if(SameSubtree(pRoot1, pRoot2)) return true;
        return HasSubtree(pRoot1->left, pRoot2) || HasSubtree(pRoot1->right, pRoot2);
    }

};

File: Algo/LeadToOffer/17_SubTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSubStructure(self, A, B):
        """
        :type A: TreeNode
        :type B: TreeNode
        :rtype: bool
        """
        if B is None and A is None:
            return True
        if B is None or A is None:
            return False

        res = False
        if A.val == B.val:
            res = self.isSameTree(A, B)
        if not res:  # you still need to check left and right
            res = self.isSubStructure(A.left, B) or self.isSubStructure(A.right, B)

        return res

    def isSameTree(self, A, B):
        if B is None:  # if the parent node are the same
            return True
        if A is None:
            return False

        if A.val != B.val:
            return False

        return self.isSameTree(A.left, B.left) and self.isSameTree(A.right, B.right)


